src/wp-admin/includes/image.php

```diff
@@ -449,24 +571,155 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
+			}
+			wp_update_attachment_metadata( $attachment_id, $image_meta );
+		}
+	}
+
+	return $image_meta;
+}
+
+/**
+ * Get the list of image sizes that support multi-mime type output.
+ *
+ * @since 6.1.0
+ *
+ * @param int $attachment_id Attachment ID.
+ * @return array the list of size names that support multi-mime type output.
+ */
+function _wp_multi_mime_get_supported_sizes( $attachment_id ) {
+	// Include only the core sizes that do not rely on add_image_size(). Additional image sizes are opt-in.
+	$default_core_sizes = array(
+		'thumbnail',
+		'medium',
+		'medium_large',
+		'large',
+		'post-thumbnail',
+	);
+
+	/**
+	 * Filter the sizes that support multi-mime type output. Developers can use this
+	 * to control the output of additional mime type sub-sized images.
+	 *
+	 * @since 6.1.0
+	 *
+	 * @param array $default_core_sizes The list of sizes that support multi-mime type output.
+	 * @param int   $attachment_id      Attachment ID.
+	 */
+	return apply_filters( 'wp_image_sizes_with_additional_mime_type_support', $default_core_sizes, $attachment_id );
+}
+
+/**
+ * Low-level function to create full-size images in additional mime types.
+ *
+ * Updates the image meta after each mime type image is created.
+ *
+ * @since 6.1.0
+ * @access private
+ *
+ * @param array  $new_mime_types Array defining what mime types to create.
+ * @param string $file           Full path to the image file.
+ * @param array  $image_meta     The attachment meta data array.
+ * @param int    $attachment_id  Attachment ID to process.
+ * @return array The attachment meta data with updated `sizes` array. Includes an array of errors encountered while resizing.
+ */
+function _wp_make_additional_mime_types( $new_mime_types, $file, $image_meta, $attachment_id ) {
+	$imagesize          = array(
+		$image_meta['width'],
+		$image_meta['height'],
+	);
+	$exif_meta          = isset( $image_meta['image_meta'] ) ? $image_meta['image_meta'] : null;
+	$original_file_size = isset( $image_meta['filesize'] ) ? $image_meta['filesize'] : wp_filesize( $file );
+
+	foreach ( $new_mime_types as $mime_type ) {
+		list( $editor, $resized, $rotated ) = _wp_maybe_scale_and_rotate_image( $file, $attachment_id, $imagesize, $exif_meta, $mime_type );
+		if ( is_wp_error( $editor ) ) {
+			// The image cannot be edited.
+			continue;
+		}
+
+		$suffix    = _wp_get_image_suffix( $resized, $rotated );
+		$extension = wp_get_default_extension_for_mime_type( $mime_type );
+		$save_as   = $editor->generate_filename( $suffix, null, $extension );
+
+		if ( file_exists( $file ) ) {
+			$save_as = $editor->generate_filename( $suffix, null, $extension, true );
+		}
+
+		$saved = $editor->save( $save_as, $mime_type );
+		if ( is_wp_error( $saved ) ) {
+			// TODO: Log errors.
+		} else {
+			// If the saved image is larger than the original, discard it.
+			$filesize = isset( $saved['filesize'] ) ? $saved['filesize'] : wp_filesize( $saved['path'] );
+			if ( $filesize > $original_file_size ) {
```

Comment:
When we do this, we must always ensure neither filesize is 0.
```suggestion
			if ( $filesize && $original_file_size && $filesize > $original_file_size ) {
```